### PR TITLE
Generalise EM algorithms

### PR DESCRIPTION
**Pull request by nguigs**
WIP
This PR aims at making the current implementation of the EM algorithm more general, in order to make it work on other manifolds.
- [x] generalise `dist_broadcast` to all metrics
- [x] move `gmm_pdf` and `weighted_gmm_pdf` to the EM module
- [x]  move `normalization_factor_init`, `find_normalization_factor`, `find_variance_from_index` to the EM module
- [x] Implement the normalisation constant for the hypersphere and other (symmetric) spaces.
- [x] generalise and rename frechet-poincarre-ball method to point_type=matrix

### CODE REVIEW

File: examples/plot_expectation_maximization_ball.py [hunk 1]
```diff
@@ -14,7 +14,8 @@
 
 import geomstats.backend as gs
 from geomstats.geometry.poincare_ball import PoincareBall
-from geomstats.learning.expectation_maximization import RiemannianEM
+from geomstats.learning.expectation_maximization import RiemannianEM, \
+    weighted_gmm_pdf
```
Comment (ninamiolane) [inline examples/plot_expectation_maximization_ball.py:18]: Should `weighted_gmm_pdf` go in the `sampling` module?

File: geomstats/_backend/numpy/__init__.py [hunk 2]
```diff
@@ -368,14 +370,14 @@ def array_from_sparse(indices, data, target_shape):
...
 
-def erf(x):
-    cst_erf = 8.0 / (3.0 * np.pi) * (np.pi - 3.0) / (4.0 - np.pi)
-    return \
-        np.sign(x) * \
-        np.sqrt(1 - np.exp(-x * x *
-                           (4 / np.pi + cst_erf * x * x) /
-                           (1 + cst_erf * x * x)))
-
+# def erf(x):
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/__init__.py:373]: leftover?

File: geomstats/_backend/tensorflow/__init__.py [hunk 3]
```diff
@@ -78,6 +78,7 @@
 cross = tf.linalg.cross
 erf = tf.math.erf
 isnan = tf.math.is_nan
+imag = tf.math.imag
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:81]: NIT: imag before isnan

File: geomstats/geometry/hypersphere.py [hunk 4]
```diff
@@ -737,6 +737,98 @@ def curvature(
...
+            sum_term = gs.sum(
+                gs.stack([summand(k)] for k in range(half_dim_2)))
+        else:
+            sum_term = summand(0)
+        coef = area * (- 1.) ** half_dim_2 / 2 ** (dim - 2) * gs.sqrt(
+            gs.pi / 2 / variances)
+
+        return coef * sum_term
+
+    def normalization_factor(self, variances):
```
Comment (ninamiolane) [inline geomstats/geometry/hypersphere.py:795]: Maybe call it `norm_factor` everywhere, since this is what is used in the name of the gradient?

File: geomstats/geometry/riemannian_metric.py [hunk 5]
```diff
@@ -290,6 +290,53 @@ def dist(self, point_a, point_b, **kwargs):
         dist = gs.sqrt(sq_dist)
         return dist
 
+    def dist_broadcast(self, point_a, point_b):
```
Comment (ninamiolane) [inline geomstats/geometry/riemannian_metric.py:293]: Why not just including this in the `dist` function?

**Review by ninamiolane — COMMENTED** [review]
Very nice until now!! Thanks for doing this! 🎉

File: examples/plot_expectation_maximization_ball.py (hunk 1 above)
Comment (nguigs) [reply examples/plot_expectation_maximization_ball.py:18]: It does not sample points from a distribution but evaluates the density function of given points. If the sampling module intends to follow tensorflow/pytorch distribution modules then yes

File: geomstats/geometry/hypersphere.py (hunk 4 above)
Comment (nguigs) [reply geomstats/geometry/hypersphere.py:795]: But it's less explicit as it can be confused with the norm of the metric

File: geomstats/geometry/riemannian_metric.py (hunk 5 above)
Comment (nguigs) [reply geomstats/geometry/riemannian_metric.py:293]: To keep things simple: the output of `dist` is a scalar or 1d, where as `dist_broadcast` is always 2d. This would be more prone to errors
